### PR TITLE
LLM: update moe block convert to optimize rest token latency of Mixtral

### PR DESCRIPTION
**Pull request by rnwang04**
### 1. Why the change?

https://github.com/analytics-zoo/nano/issues/847#issuecomment-1851572877
https://github.com/analytics-zoo/nano/issues/847#issuecomment-1853158787

### 2. User API changes

No change. Just set `optimize_model=True` for Mixtral.

### 3. Summary of the change 

- convert `MixtralSparseMoeBlock` of mixtral to simpify rest token inference logic, which greatly reduce latency

performance and accuracy validation is here: https://github.com/analytics-zoo/nano/issues/847#issuecomment-1853368957

### 4. How to test?
- [x] Local test with `Mixtral-8x7B-Instruct-v0.1`
- [x] Unit test

### CODE REVIEW

File: python/llm/src/bigdl/llm/transformers/models/mixtral.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+        )
+        # One hot encode the selected experts to create an expert mask
+        # this will be used to easily index which expert is going to be sollicitated
+        expert_mask = torch.nn.functional.one_hot(selected_experts,
+                                                  num_classes=self.num_experts).permute(2, 1, 0)
+
+        # Loop over all available experts in the model and perform the computation on each expert
+        for expert_idx in range(self.num_experts):
+            expert_layer = self.experts[expert_idx]
+            idx, top_x = torch.where(expert_mask[expert_idx])
```
Comment (shane-huang) [inline python/llm/src/bigdl/llm/transformers/models/mixtral.py:77]: shall we also optimize this part (e.g. replace w/ c++ implementation as yang suggested?) for the 1st token speed?

**Review by shane-huang — APPROVED** [review]

File: python/llm/src/bigdl/llm/transformers/models/mixtral.py (hunk 1 above)
Comment (rnwang04) [reply python/llm/src/bigdl/llm/transformers/models/mixtral.py:77]: for first token optimization, may need further validation, maybe we can optimize this part later if needed.

Comment (jason-dai) [reply python/llm/src/bigdl/llm/transformers/models/mixtral.py:77]: We also need to optimize sequence_length>1 and batch_size>1 for finetuning
